File: scripts/validate-changelog.rb
#!/usr/bin/env ruby

require "date"

if ARGV.length > 1
  warn "Usage: #{File.basename($PROGRAM_NAME)} [repository root]"
  exit 2
end

repository_root = File.expand_path(ARGV.fetch(0, "."))
unless Dir.exist?(repository_root)
  warn "Changelog check: repository root does not exist: #{repository_root}"
  exit 2
end

changelog_path = File.join(repository_root, "CHANGELOG.md")
unless File.file?(changelog_path)
  warn "Changelog check: missing CHANGELOG.md"
  exit 2
end

lines = File.readlines(changelog_path, chomp: true)
errors = []

errors << "the first heading must be '# Changelog'" unless lines.first == "# Changelog"

allowed_categories = %w[Added Changed Deprecated Removed Fixed Security]
semver_pattern = /\A(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|[0-9A-Za-z-]*[A-Za-z-][0-9A-Za-z-]*)(?:\.(?:0|[1-9]\d*|[0-9A-Za-z-]*[A-Za-z-][0-9A-Za-z-]*))*))?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?\z/
sections = []
current_section = nil
current_category = nil
category_has_entry = false

record_category = lambda do
  if current_category && !category_has_entry
    section_label = current_section ? current_section[:label] : "an invalid section"
    errors << "category #{current_category.inspect} in #{section_label} has no entry"
  end
end

lines.each_with_index do |line, index|
  line_number = index + 1

  if line.match?(/\A## /)
    record_category.call
    current_category = nil
    category_has_entry = false

    match = line.match(/\A## \[([^\]]+)\](?: - (\d{4}-\d{2}-\d{2}))?\z/)
    unless match
      errors << "line #{line_number} has an invalid section heading"
      current_section = nil
      next
    end

    label = match[1]
    date_text = match[2]
    if label == "Unreleased"
      errors << "Unreleased must not have a date" if date_text
    elsif date_text.nil? || !semver_pattern.match?(label)
      errors << "release section #{label.inspect} must use SemVer and an ISO date"
    end

    release_date = nil
    if date_text
      begin
        release_date = Date.iso8601(date_text)
      rescue ArgumentError
        errors << "release section #{label.inspect} has an invalid ISO date"
      end
    end

    current_section = { label: label, date: release_date, version: label == "Unreleased" ? nil : label }
    sections << current_section
    next
  end

  if line.match?(/\A### /)
    record_category.call
    category = line.delete_prefix("### ")
    if current_section.nil?
      errors << "line #{line_number} has a category outside a changelog section"
    elsif !allowed_categories.include?(category)
      errors << "line #{line_number} uses unsupported category #{category.inspect}"
    end
    current_category = category
    category_has_entry = false
    next
  end

  if line.start_with?("- ") && current_section && current_category && !line.delete_prefix("- ").strip.empty?
    category_has_entry = true
  end
end

record_category.call

unreleased_sections = sections.select { |section| section[:label] == "Unreleased" }
errors << "the changelog must contain exactly one [Unreleased] section" unless unreleased_sections.length == 1
if sections.any? && sections.first[:label] != "Unreleased"
  errors << "[Unreleased] must be the first changelog section"
end

release_sections = sections.reject { |section| section[:label] == "Unreleased" }
if release_sections.any? { |section| section[:date].nil? || section[:version].nil? }
  errors << "every release section must have a valid date and version"
end

if release_sections.map { |section| section[:version] }.compact.uniq.length != release_sections.length
  errors << "release versions must be unique"
end

def compare_versions(left, right)
  left_match = left.match(/\A(\d+)\.(\d+)\.(\d+)(?:-([^+]+))?/)
  right_match = right.match(/\A(\d+)\.(\d+)\.(\d+)(?:-([^+]+))?/)
  left_core = left_match[1, 3].map(&:to_i)
  right_core = right_match[1, 3].map(&:to_i)
  core_comparison = left_core <=> right_core
  return core_comparison unless core_comparison.zero?

  left_pre = left_match[4]&.split(".")
  right_pre = right_match[4]&.split(".")
  return 0 if left_pre.nil? && right_pre.nil?
  return 1 if left_pre.nil?
  return -1 if right_pre.nil?

  left_pre.zip(right_pre).each do |left_id, right_id|
    return -1 if left_id.nil?
    return 1 if right_id.nil?
    if left_id.match?(/\A\d+\z/) && right_id.match?(/\A\d+\z/)
      comparison = left_id.to_i <=> right_id.to_i
    elsif left_id.match?(/\A\d+\z/)
      comparison = -1
    elsif right_id.match?(/\A\d+\z/)
      comparison = 1
    else
      comparison = left_id <=> right_id
    end
    return comparison unless comparison.zero?
  end
  0
end

valid_release_sections = release_sections.select do |section|
  section[:date] && section[:version] && semver_pattern.match?(section[:version])
end

valid_release_sections.each_cons(2) do |newer, older|
  if compare_versions(newer[:version], older[:version]) <= 0
    errors << "release sections must be ordered by descending SemVer"
  end
  if newer[:date] && older[:date] && newer[:date] < older[:date]
    errors << "release sections must be ordered by descending date"
  end
end

if errors.empty?
  puts "Changelog check passed."
  exit 0
end

errors.each { |error| warn "Changelog check: #{error}" }
warn "Changelog check failed with #{errors.length} error(s)."
exit 1

File: scripts/validate-changelog.mjs
// agentic-primitive: {"id":"changelog-structure-validator","kind":"validator","enforcement":"deterministic","adrs":["ADR-0006"],"domains":["agentic-delivery-governance"]}
import { access, readFile, stat } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const allowedCategories = new Set(['Added', 'Changed', 'Deprecated', 'Removed', 'Fixed', 'Security']);
const semverPattern = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|[0-9A-Za-z-]*[A-Za-z-][0-9A-Za-z-]*)(?:\.(?:0|[1-9]\d*|[0-9A-Za-z-]*[A-Za-z-][0-9A-Za-z-]*))*))?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/;

export class ChangelogValidationError extends Error {
  constructor(message, exitCode = 1) {
    super(message);
    this.name = 'ChangelogValidationError';
    this.exitCode = exitCode;
  }
}

async function isDirectory(filePath) {
  try {
    return (await stat(filePath)).isDirectory();
  } catch {
    return false;
  }
}

async function isFile(filePath) {
  try {
    return (await stat(filePath)).isFile();
  } catch {
    return false;
  }
}

function parseIsoDate(value) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return null;
  const [year, month, day] = value.split('-').map(Number);
  const date = new Date(`${value}T00:00:00.000Z`);
  if (Number.isNaN(date.getTime())) return null;
  if (date.getUTCFullYear() !== year || date.getUTCMonth() + 1 !== month || date.getUTCDate() !== day) {
    return null;
  }
  return value;
}

function compareVersions(left, right) {
  const leftMatch = left.match(/^(\d+)\.(\d+)\.(\d+)(?:-([^+]+))?/);
  const rightMatch = right.match(/^(\d+)\.(\d+)\.(\d+)(?:-([^+]+))?/);
  const leftCore = leftMatch.slice(1, 4).map(Number);
  const rightCore = rightMatch.slice(1, 4).map(Number);
  for (let index = 0; index < leftCore.length; index += 1) {
    if (leftCore[index] !== rightCore[index]) return leftCore[index] - rightCore[index];
  }

  const leftPre = leftMatch[4]?.split('.');
  const rightPre = rightMatch[4]?.split('.');
  if (!leftPre && !rightPre) return 0;
  if (!leftPre) return 1;
  if (!rightPre) return -1;
  for (let index = 0; index < Math.max(leftPre.length, rightPre.length); index += 1) {
    const leftId = leftPre[index];
    const rightId = rightPre[index];
    if (leftId === undefined) return -1;
    if (rightId === undefined) return 1;
    const leftNumeric = /^\d+$/.test(leftId);
    const rightNumeric = /^\d+$/.test(rightId);
    if (leftNumeric && rightNumeric) {
      if (Number(leftId) !== Number(rightId)) return Number(leftId) - Number(rightId);
    } else if (leftNumeric !== rightNumeric) {
      return leftNumeric ? -1 : 1;
    } else if (leftId !== rightId) {
      return leftId < rightId ? -1 : 1;
    }
  }
  return 0;
}

export async function validateChangelog(repositoryRoot = process.cwd()) {
  const root = path.resolve(repositoryRoot);
  if (!(await isDirectory(root))) {
    throw new ChangelogValidationError(`Changelog check: repository root does not exist: ${repositoryRoot}`, 2);
  }
  const changelogPath = path.join(root, 'CHANGELOG.md');
  if (!(await isFile(changelogPath))) {
    throw new ChangelogValidationError('Changelog check: missing CHANGELOG.md', 2);
  }

  const lines = (await readFile(changelogPath, 'utf8')).split(/\r?\n/);
  if (lines.at(-1) === '') lines.pop();
  const errors = [];
  if (lines[0] !== '# Changelog') errors.push("the first heading must be '# Changelog'");

  const sections = [];
  let currentSection = null;
  let currentCategory = null;
  let categoryHasEntry = false;

  const recordCategory = () => {
    if (currentCategory && !categoryHasEntry) {
      const sectionLabel = currentSection?.label ?? 'an invalid section';
      errors.push(`category ${JSON.stringify(currentCategory)} in ${sectionLabel} has no entry`);
    }
  };

  lines.forEach((line, index) => {
    const lineNumber = index + 1;
    if (/^## /.test(line)) {
      recordCategory();
      currentCategory = null;
      categoryHasEntry = false;
      const match = line.match(/^## \[([^\]]+)\](?: - (\d{4}-\d{2}-\d{2}))?$/);
      if (!match) {
        errors.push(`line ${lineNumber} has an invalid section heading`);
        currentSection = null;
        return;
      }
      const label = match[1];
      const dateText = match[2];
      if (label === 'Unreleased') {
        if (dateText) errors.push('Unreleased must not have a date');
      } else if (!dateText || !semverPattern.test(label)) {
        errors.push(`release section ${JSON.stringify(label)} must use SemVer and an ISO date`);
      }
      const releaseDate = dateText ? parseIsoDate(dateText) : null;
      if (dateText && !releaseDate) errors.push(`release section ${JSON.stringify(label)} has an invalid ISO date`);
      currentSection = { label, date: releaseDate, version: label === 'Unreleased' ? null : label };
      sections.push(currentSection);
      return;
    }
    if (/^### /.test(line)) {
      recordCategory();
      const category = line.slice(4);
      if (!currentSection) {
        errors.push(`line ${lineNumber} has a category outside a changelog section`);
      } else if (!allowedCategories.has(category)) {
        errors.push(`line ${lineNumber} uses unsupported category ${JSON.stringify(category)}`);
      }
      currentCategory = category;
      categoryHasEntry = false;
      return;
    }
    if (line.startsWith('- ') && currentSection && currentCategory && line.slice(2).trim()) {
      categoryHasEntry = true;
    }
  });
  recordCategory();

  const unreleasedSections = sections.filter(({ label }) => label === 'Unreleased');
  if (unreleasedSections.length !== 1) errors.push('the changelog must contain exactly one [Unreleased] section');
  if (sections.length > 0 && sections[0].label !== 'Unreleased') errors.push('[Unreleased] must be the first changelog section');

  const releaseSections = sections.filter(({ label }) => label !== 'Unreleased');
  if (releaseSections.some(({ date, version }) => !date || !version)) {
    errors.push('every release section must have a valid date and version');
  }
  if (new Set(releaseSections.map(({ version }) => version)).size !== releaseSections.length) {
    errors.push('release versions must be unique');
  }

  const validReleaseSections = releaseSections.filter(({ date, version }) => date && version && semverPattern.test(version));
  for (let index = 0; index < validReleaseSections.length - 1; index += 1) {
    const newer = validReleaseSections[index];
    const older = validReleaseSections[index + 1];
    if (compareVersions(newer.version, older.version) <= 0) errors.push('release sections must be ordered by descending SemVer');
    if (newer.date < older.date) errors.push('release sections must be ordered by descending date');
  }

  if (errors.length > 0) {
    throw new ChangelogValidationError(
      `${errors.map((error) => `Changelog check: ${error}`).join('\n')}\nChangelog check failed with ${errors.length} error(s).`,
      1,
    );
  }
}

const isMainModule = process.argv[1]
  && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url);

if (isMainModule) {
  if (process.argv.length > 3) {
    process.stderr.write(`Usage: ${path.basename(process.argv[1])} [repository root]\n`);
    process.exitCode = 2;
  } else {
    try {
      await validateChangelog(process.argv[2] ?? '.');
      process.stdout.write('Changelog check passed.\n');
    } catch (error) {
      process.stderr.write(`${error.message}\n`);
      process.exitCode = error.exitCode ?? 1;
    }
  }
}
